### PR TITLE
Improve recipient bar layout and design

### PR DESCRIPTION
**Pull request by sumanthvrao**
This involves redoing the top bar, to better represent the current narrow and remove redundant information.
*All message view* - 

all_msg

*Stream view* -
stream

Currently it supports for :
* All messages,
* All Private messages,
* Streams
* Stream and topic
* PM's
* Group PM's
* Starred messages

Fixes: #340 
**Based on discussion in** - https://chat.zulip.org/#narrow/stream/206-zulip-terminal/topic/top.20recipient.20bar/near/733211

### CODE REVIEW

**Comment by sumanthvrao** [comment]
This is ready to be taken out on a drive :). Tests are yet to be added. Love to hear your feedback. 
(@neiljp @amanagr @rishig)

File: zulipterminal/ui_tools/boxes.py [hunk 1]
```diff
@@ -201,6 +201,46 @@ def stream_header(self) -> Any:
...
+        elif self.message['type'] == 'stream':
+            bar_color = self.model.stream_dict[self.stream_id]['color']
+            bar_color = 's' + bar_color[:2] + bar_color[3] + bar_color[5]
+            text_to_fill = ('bar', [  # type: ignore
+                (bar_color, ' {} '.format(self.caption))
+            ])
+        else:
+            text_to_fill = 'Private message'
+        title_markup = ('header', [
+            ('custom', 'Conversation focus'),
```
Comment (neiljp) [inline zulipterminal/ui_tools/boxes.py]: I recall suggesting this name, but perhaps we can move whatever text we decide on for this to the title for the frame, so it appears above the box instead?

Comment (sumanthvrao) [reply zulipterminal/ui_tools/boxes.py]: Makes sense. Have moved it above.

**Comment by sumanthvrao** [comment]
@neiljp Thanks for the review :+1: 

* The first two points - of renaming 'Messages' and moving 'conversation focus'  - I did the changes accordingly. We can change the names of these to an apt one based on other suggestions.
* With regard to `z` on narrows. I am not understanding what you mean her. The narrow **does indeed** change for me from `All private messages` to `Private messages` works. But as you say for the `streams` there is redundant information in the 2 bars. Will see what I can do here.

**Comment by amanagr** [comment]
Great work @sumanthvrao! Looks nice and fine to me. You can start working on tests for the changes.

**Comment by sumanthvrao** [comment]
Thanks for the praise @amanagr! Yup I am adding tests :)

File: zulipterminal/ui_tools/boxes.py [hunk 2]
```diff
@@ -525,17 +525,26 @@ def main_view(self) -> Any:
...
             ('pack', urwid.Text("  ")),
             self.text_box,
         ])
         self.w = urwid.LineBox(
             w, tlcorner=u'', tline=u'', lline=u'',
-            trcorner=u'', blcorner=u'─', rline=u'',
-            bline=u'─', brcorner=u'─')
-        return [self.w]
+            trcorner=u'', blcorner=u'', rline=u'',
+            bline=u'', brcorner=u'')
```
Comment (neiljp) [inline zulipterminal/ui_tools/boxes.py]: Does this `LineBox` actually do anything?

File: zulipterminal/ui_tools/boxes.py [hunk 3]
```diff
@@ -163,6 +163,15 @@ def _time_for_message(self, message: Dict[str, Any]) -> str:
...
 
     def need_recipient_header(self) -> bool:
+        # Prevent redundant information in recipient bar
+        # and recipient header.
+        if len(self.model.narrow) == 1 and \
+                self.model.narrow[0][0] == 'pm_with':
+            return False
+        if len(self.model.narrow) == 2 and \
+                self.model.narrow[1][0] == 'topic':
+            return False
```
Comment (neiljp) [inline zulipterminal/ui_tools/boxes.py]: I see what you've achieved with this commit and I think it's useful.
Does this just pass an existing test? You've not added a test to check this directly?

I wanted to clarify that this wasn't quite what I meant with hiding the recipient when it's "fully narrowed"; in that case we can of course do something like you have done here, but also we can merge the two boxes together (or hide the bottom one, perhaps).

File: zulipterminal/ui_tools/boxes.py (hunk 2 above)
Comment (sumanthvrao) [reply zulipterminal/ui_tools/boxes.py]: Ah, I see what you mean here. Yup. This does not seem to add anything. Will remove the `LineBox` in the next iteration.

File: zulipterminal/ui_tools/boxes.py (hunk 3 above)
Comment (sumanthvrao) [reply zulipterminal/ui_tools/boxes.py]: This does pass the existing test. However I did not find any tests specifically for this. 
 
Yup. I understood what you mean earlier. We can make this a v2 if you agree, and merge the two boxes together, in specific narrows.

Comment (sumanthvrao) [reply zulipterminal/ui_tools/boxes.py]: @neiljp Do you want me to add tests for this? There are no previous tests for `need_recipient_header` If I am not wrong.

**Comment by sumanthvrao** [comment]
[Floating Recipient Bar](https://trello.com/c/IKc5KuY8/2-floating-recipient-bar)

File: zulipterminal/ui_tools/boxes.py (hunk 3 above)
Comment (neiljp) [reply zulipterminal/ui_tools/boxes.py]: Agreed; when merged it will be easier to get feedback :)

**Comment by sumanthvrao** [comment]
Rebased this PR with some changes from previous iterations. 
The major one being, splitting the second commit into - one that build the 2 headers and one that updates them. Tests have been added for the same.
@neiljp @amanagr Its now ready for review/feedback :)

File: zulipterminal/ui_tools/boxes.py [hunk 4]
```diff
@@ -219,6 +219,47 @@ def private_header(self) -> Any:
...
+                self.model.narrow[0][1] == 'starred':
+            text_to_fill = 'Starred messages'
+        elif self.message['type'] == 'stream':
+            bar_color = self.model.stream_dict[self.stream_id]['color']
+            bar_color = 's' + bar_color[:2] + bar_color[3] + bar_color[5]
+            if len(self.model.narrow) == 2 and \
+                    self.model.narrow[1][0] == 'topic':
+                text_to_fill = ('bar', [  # type: ignore
+                    (bar_color, '{}'.format(self.caption)),
+                    (bar_color, ': topic narrow ')
```
Comment (neiljp) [inline zulipterminal/ui_tools/boxes.py]: I'm not too keen on this, though unless we collapse the bar below I'm not sure what to do other than use a better phrase.

However, there's an extra space at the end here? That would probably be clearer if the tests were more complete?

File: zulipterminal/ui_tools/boxes.py [hunk 5]
```diff
@@ -541,17 +541,21 @@ def main_view(self) -> Any:
         self.text_box = ReadlineEdit(u"Search: ")
         # Add some text so that when packing,
         # urwid doesn't hide the widget.
-        self.msg_narrow = urwid.Text("DONT HIDE")
-        w = urwid.Columns([
-            ('pack', self.msg_narrow),
+        self.conversation_focus = urwid.Text("DONT HIDE")
```
[the remark this answers is not in the document]
Comment (sumanthvrao) [reply zulipterminal/ui_tools/boxes.py]: Thanks for notifying this. I did not in-fact know about #201. Have changed the commit and commit summary to appear as a fix to that issue.

File: zulipterminal/ui_tools/boxes.py [hunk 6]
```diff
@@ -219,6 +219,47 @@ def private_header(self) -> Any:
...
         return header
 
+    def top_header_bar(self, message_view: Any) -> Any:
+        if self.message['type'] == 'stream':
+            return message_view.stream_header()
+        else:
+            return message_view.private_header()
+
+    def top_search_bar(self) -> Any:
+        if self.model.narrow == []:
```
[the remark this answers is not in the document]
Comment (sumanthvrao) [reply zulipterminal/ui_tools/boxes.py]: Done :+1:

File: zulipterminal/ui_tools/boxes.py [hunk 7]
```diff
@@ -219,6 +219,47 @@ def private_header(self) -> Any:
...
+                ])
+            else:
+                text_to_fill = ('bar', [  # type: ignore
+                    (bar_color, '{}'.format(self.caption))
+                ])
+        elif len(self.model.narrow) == 1 and \
+                len(self.model.narrow[0][1].split(",")) > 1:
+            text_to_fill = 'Group private message'
+        else:
+            text_to_fill = 'Private message'
```
[the remark this answers is not in the document]
Comment (sumanthvrao) [reply zulipterminal/ui_tools/boxes.py]: Re 'Private' : We could definitely skip the 'private' part here. I understand why. How about we get some feedback before that?
Re 'distinguish' : I agree about this. Well I kind off liked the distinction. The earlier suggestion of yours, 'just merging the 2 bars in small narrow' would certainly avoid this confusion (v2?)

File: zulipterminal/ui_tools/boxes.py (hunk 4 above)
Comment (sumanthvrao) [reply zulipterminal/ui_tools/boxes.py]: We could go for a better phrase. :)
Will amend tests. I will definitely do a v2 of this shortly, - merge the two bars in smaller narrows.

File: tests/ui/test_ui_tools.py [hunk 8]
```diff
@@ -1114,6 +1114,41 @@ def test_main_view_generates_PM_header(self, mocker, message,
...
+        header_bar = msg_box.top_header_bar(msg_box)
+
+        if current_message['type'] == 'stream':
+            assert header_bar.text.startswith('PTEST >')
+        else:
+            assert header_bar.text.startswith('Private Messages with: ')
+        if len(msg_narrow) > 0 and msg_narrow[0][0] == 'stream':
+            assert len(search_bar.markup[1][0][1][1]) == assert_narrow
+        else:
+            assert search_bar.markup[1][0][1] == assert_narrow
```
[the remark this answers is not in the document]
Comment (sumanthvrao) [reply tests/ui/test_ui_tools.py]: Agreed! I in-fact don't know why I made the tests so complex to start with. These have been fixed simplified. (As you said, slightly less coverage seems to be the side effect here)

**Comment by sumanthvrao** [comment]
@neiljp updated this PR. The major changes from previous iteration involved making the tests more general.

File: zulipterminal/ui_tools/boxes.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (neiljp) [reply zulipterminal/ui_tools/boxes.py]: This looks more readable now; you agree?

[the remark this answers is not in the document]
Comment (sumanthvrao) [reply zulipterminal/ui_tools/boxes.py]: Definitely! Agreed :)

**Comment by neiljp** [comment]
@sumanthvrao Thanks, I think this is great and hopefully others will too! :tada:

As you know, the first commit was merged separately, and then I increased the number of test cases in the remaining commits, before getting your approval and merging the others.

**Comment by sumanthvrao** [comment]
@neiljp Thanks for adding the additional tests. Glad to have this merged :)

**Pull request closed without merging**